File: protocol/bet.py
class Bet:
    def __init__(self, nombre, apellido, dni, nacimiento, numero):
        self.nombre = nombre
        self.apellido = apellido
        self.dni = dni
        self.nacimiento = nacimiento
        self.numero = numero

    def to_bytes(self):
        """Serializes the Bet object to a byte array."""
        data = bytearray()
        
        first_name_bytes = self.nombre.encode('utf-8')
        data.extend(len(first_name_bytes).to_bytes(1, byteorder='big'))
        data.extend(first_name_bytes)
        
        last_name_bytes = self.apellido.encode('utf-8')
        data.extend(len(last_name_bytes).to_bytes(1, byteorder='big'))
        data.extend(last_name_bytes)
        
        data.extend(self.dni.to_bytes(4, byteorder='big'))

        year = self.nacimiento.split("-")[0]
        month = self.nacimiento.split("-")[1]
        day = self.nacimiento.split("-")[2]

        data.extend(int(year).to_bytes(2, byteorder='big'))
        data.extend(int(month).to_bytes(1, byteorder='big'))
        data.extend(int(day).to_bytes(1, byteorder='big'))
        
        data.extend(self.numero.to_bytes(4, byteorder='big'))

        return bytes(len(data).to_bytes(2, byteorder='big') + data)
    
    @classmethod
    def from_bytes(self, data):
        """Deserializes a byte array to a Bet object."""
        index = 0

        first_name_length = data[index]
        index += 1
        first_name = data[index:index + first_name_length].decode('utf-8')
        index += first_name_length
        
        last_name_length = data[index]
        index += 1
        last_name = data[index:index + last_name_length].decode('utf-8')
        index += last_name_length

        dni = int.from_bytes(data[index:index + 4], byteorder='big')
        index += 4

        year = int.from_bytes(data[index:index + 2], byteorder='big')
        index += 2
        month = int.from_bytes(data[index:index + 1], byteorder='big')
        index += 1
        day = int.from_bytes(data[index:index + 1], byteorder='big')
        index += 1

        birth_date = f"{year}-{month:02d}-{day:02d}"

        number = int.from_bytes(data[index:index + 4], byteorder='big')
        
        return self(first_name, last_name, dni, birth_date, number)
    
    def __repr__(self):
        return (f"Bet(nombre={self.nombre}, apellido={self.apellido}, dni={self.dni}, "
                f"nacimiento={self.nacimiento}, numero={self.numero})")
    

bet = Bet("Juan", "Perez", 12345678, "1990-01-01", 42)
print("bet original", bet)

bytes = bet.to_bytes()

len = int.from_bytes(bytes[:2], byteorder='big')

bet_recv = Bet.from_bytes(bytes[2:len + 2])

print("bet_recv", bet_recv)

File: protocol/bet.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type Bet struct {
	Nombre    string
	Apellido  string
	DNI       uint32
	Nacimiento string
	Numero    uint32
}

func (b *Bet) toBytes() []byte {
	var data []byte

	firstNameBytes := []byte(b.Nombre)
	data = append(data, byte(len(firstNameBytes)))
	data = append(data, firstNameBytes...)

	lastNameBytes := []byte(b.Apellido)
	data = append(data, byte(len(lastNameBytes)))
	data = append(data, lastNameBytes...)

	data = append(data, byte(b.DNI>>24), byte(b.DNI>>16), byte(b.DNI>>8), byte(b.DNI))

	dateParts := strings.Split(b.Nacimiento, "-")
	year, _ := strconv.Atoi(dateParts[0])
	month, _ := strconv.Atoi(dateParts[1])
	day, _ := strconv.Atoi(dateParts[2])

	data = append(data, byte(year>>8), byte(year))
	data = append(data, byte(month))
	data = append(data, byte(day))

	data = append(data, byte(b.Numero>>24), byte(b.Numero>>16), byte(b.Numero>>8), byte(b.Numero))

	return append([]byte{byte(len(data) >> 8), byte(len(data))}, data...)
}

func fromBytes(data []byte) *Bet {
	index := 0

	firstNameLength := int(data[index])
	index++
	firstName := string(data[index : index+firstNameLength])
	index += firstNameLength

	lastNameLength := int(data[index])
	index++
	lastName := string(data[index : index+lastNameLength])
	index += lastNameLength

	dni := uint32(data[index])<<24 | uint32(data[index+1])<<16 | uint32(data[index+2])<<8 | uint32(data[index+3])
	index += 4

	year := int(data[index])<<8 | int(data[index+1])
	index += 2
	month := int(data[index])
	index++
	day := int(data[index])
	index++

	birthDate := fmt.Sprintf("%04d-%02d-%02d", year, month, day)

	number := uint32(data[index])<<24 | uint32(data[index+1])<<16 | uint32(data[index+2])<<8 | uint32(data[index+3])

	return &Bet{
		Nombre:    firstName,
		Apellido:  lastName,
		DNI:       dni,
		Nacimiento: birthDate,
		Numero:    number,
	}
}

func main() {
	bet := Bet{"Juan", "Perez", 12345678, "1990-01-01", 42}
	fmt.Println("bet original", bet)

	bytes := bet.toBytes()
	fmt.Printf("bytes %v\n", bytes)

	length := int(bytes[0])<<8 | int(bytes[1])
	betRecv := fromBytes(bytes[2:length+2])

	fmt.Println("bet_recv", *betRecv)
}